### PR TITLE
Comment enhancements

### LINKED ISSUE #27: Multi-line XML doc comments
Example:

```C#
\**
<summary>Hello!</summary>
*\
class C { }
```

### LINKED ISSUE #80: Syntax highlighting lost after certain comments
_From @cateyes99 on July 23, 2017 3:27_

## Environment data
`dotnet --info` output:
```
.NET Command Line Tools (1.0.1)

Product Information:
 Version:            1.0.1
 Commit SHA-1 hash:  005db40cd1

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.11
 OS Platform: Darwin
 RID:         osx.10.11-x64
 Base Path:   /usr/local/share/dotnet/sdk/1.0.1
```
VS Code version: `1.14.2`
C# Extension version: `1.11.0`

## Steps to reproduce
Create below C# snippet
image

## Expected  behavior
The method `F2()` should still have correct syntax highlighting.

## Actual behavior
The method `F2()` doesn't have correct syntax highlighting.

I **notice**:
if I remove the **Line 9**, the the syntax highlighting is back:
image

or remove the **namespance**, the the syntax highlighting is also back:
image

_Copied from original issue: OmniSharp/omnisharp-vscode#1659_

### LINKED ISSUE #151: Highlight delimited documentation comments /** */
> Issue Type: Feature Request
> 
> According to the C♯ specification, there's a [delimited documentation comment syntax](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments#introduction), and VS Code supports this (eg, Intellisense).
> However, the XML in delimited documentation comments such as
> 
> ```cs
> /**
> <summary>
> 	Filler
> </summary>
> */
> ```
> 
> is not getting syntax highlighting that single line comment syntax such as
> 
> ```cs
> /// <summary>
> /// 	Filler
> /// </summary>
> ```
> 
> is getting.
> Can this be added?
> 
> Extension version: 1.18.0-beta7
> VS Code version: Code 1.31.1 (1b8e8302e405050205e69b59abb3559592bb9e60, 2019-02-12T02:20:54.427Z)
> OS version: Windows_NT x64 10.0.17763

### LINKED ISSUE #177: C# syntax highlighting, after wrong XML Comment doesn't work.
From @saxc in https://github.com/microsoft/vscode/issues/91292

- VSCode Version: 1.43.0-insider (user setup)
- Commit: 07c33977da255861edf71e65b18249d2f67d0762
- OS Version: Windows_NT x64 10.0.18363
- without extensions

Steps to Reproduce:

The test code was write in VS 2019 (works).
By removing an quote in the XML comment, VS Code shows all subsequent lines as string.

VS-Code

```csharp
using System;

namespace ConsoleApp1
{
    class Program
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="args></param>
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }
    }
}
```

### PR DESCRIPTION
**Pull request by wise0704**
1. Delimited documentation
Resolves #151
Resolves #27
image &rArr; image

2. Line comment only consumes one line
Resolves #177 
image &rArr; image

3. Line comment completely consumes the entire line, so lookbehinds do not detect the last character of the comment
Resolves #80
image &rArr; image

4. Doc comment content across multiple lines
image &rArr; image

5. Delimited doc comment content across multiple lines **(WIP)**
image &rArr; image

### CODE REVIEW

File: src/csharp.tmLanguage.yml [hunk 1]
```diff
@@ -3420,30 +3420,39 @@ repository:
...
+      patterns:
+      - match: (^\s*)(\*)(?!/)
+        captures:
+         1: { name: punctuation.whitespace.comment.leading.cs }
+         2: { name: punctuation.definition.comment.cs }
+      - include: '#xml-doc-comment'
+    - name: comment.line.double-slash.cs
+      begin: (^\s+)?(//).*$
+      beginCaptures:
+        1: { name: punctuation.whitespace.comment.leading.cs }
```
Comment (wise0704) [inline src/csharp.tmLanguage.yml:3446]: Why do we need to highlight leading whitespace?

**Comment by wise0704** [comment]
Delimited doc seems too difficult to properly handle.

Currently:

- If leading asterisks are present, then
  - Start token does not flow to leading asterisk lines.
image
  - End token cannot be on a leading asterisk line .
image

File: src/csharp.tmLanguage.yml (hunk 1 above)
Comment (JoeRobich) [reply src/csharp.tmLanguage.yml:3449]: I don't know why. It doesn't seem useful to me. The C# grammar started as a fork of the typescript grammar, and I can see that they emit this scope as well. https://grep.app/search?q=punctuation.whitespace.comment.leading

Comment (wise0704) [reply src/csharp.tmLanguage.yml:3449]: There must be some reason. At least at the moment there's no harm in keeping it.

**Comment by wise0704** [comment]
begin/while is the only way to separate leading patterns to the content on the right, which works perfectly if all leading patterns are the same (like `>` in markdown)

But it doesn't allow any form of context closing pattern. All that's needed is some feature that allows "process until this line and stop looking into next line if this condition is met".
If such `end` condition was supported on begin/while, then a simple `end: */` would solve all the issues above, with

```yaml
# pseudocode
begin: '/**'
while: '*' ?
end: '*/'
```

Tried all forms of chaining begin/end and begin/while but fundamentally since fundamentally begin/while is meant to consume the entire line, no combination of end condition will be able to break out of it.

Things I tried, were

```yaml
begin: (?= '/**')
while: '*' ?
  begin: '/**'
    include: #xml-doc-comment
  end: '*/'
```

but the `while` condition will keep consuming consecutive lines since it does not know whether the end condition was met. Similarly

```yaml
begin: '/**'
  begin: \G
  while: '*' ?
    include: #xml-doc-comment
end: '*/'
```

will cause the while condition to keep consuming all lines, not giving end condition a chance to match. Having something like `(?=(?~\*/)$)` will make it stop, but it will stop on the line **before** the line containing the closing token. Including this line separately can be done, 

```yaml
begin: '/**'
  begin: \G (?= (?~ '*/') $)
  while: '*' ? (?= (?~ '*/') $)
    include: #xml-doc-comment
  include: #xml-doc-comment
end: '*/'
```

which is the current implementation as of this comment, but the line containing `*/` will not connect with the rest.

File: test/preprocessor.tests.ts [hunk 2]
```diff
@@ -217,25 +217,6 @@ describe("Preprocessor", () => {
             ]);
         });
 
-        it("#if (Foo != true) && Bar //Foo", async () => {
```
Comment (JoeRobich) [inline test/preprocessor.tests.ts:220]: Why was this removed? Do you not believe the test is useful or do we no longer support comments on preprocessor lines?

**Review by JoeRobich — APPROVED** [review]
@wise0704 Thanks for putting so much effort into implementing this. I've probably learned more regex and textmate in the last month from reading your PRs than I knew before.
